pkgs/racket-doc/scribblings/reference/eval-model.scrbl

```diff
@@ -60,55 +59,57 @@ Some simplifications require more than one step. For example:
 ]
 
 An expression that is not a @tech{value} can always be partitioned
-into two parts: a @deftech{redex}, which is the part that changed in a
+into two parts: a @deftech{redex} (``reducible expression''),
+which is the part that can change in a
 single-step simplification (highlighted), and the
 @deftech{continuation}, which is the evaluation
-context surrounding an expression. In @racket[(- 4 (+ 1 1))], the redex is @racket[(+ 1 1)], and
-the continuation is @racket[(- 4 @#,hole)], where @hole takes the
-place of the redex. That is, the continuation says how to ``continue''
+context surrounding the redex. In @racket[(- 4 (+ 1 1))], the redex is @racket[(+ 1 1)], and
+the continuation is @racket[(- 4 @#,hole)], where @hole indicates the
+position of the redex. That is, the continuation says how to ``continue''
 after the @tech{redex} is reduced to a @tech{value}.
 
-Before some things can be evaluated, some sub-expressions must be
-evaluated; for example, in the application @racket[(- 4 (+ 1 1))], the
+Before some expressions can be evaluated, their sub-expressions must be
+evaluated. For example, in the application @racket[(- 4 (+ 1 1))], the
 application of @racket[-] cannot be reduced until the sub-expression
 @racket[(+ 1 1)] is reduced.
-
 Thus, the specification of each syntactic form specifies how (some of)
-its sub-expressions are evaluated, and then how the results are
+its sub-expressions are evaluated and then how the results are
 combined to reduce the form away.
 
 The @deftech{dynamic extent} of an expression is the sequence of
-evaluation steps during which the expression contains the @tech{redex}.
+evaluation steps starting when the expression is selected as a @tech{redex}
+and continuing until it is fully evaluated.
 
 @;------------------------------------------------------------------------
 @section{Tail Position}
 
 An expression @racket[_expr1] is in @deftech{tail position} with
 respect to an enclosing expression @racket[_expr2] if, whenever
 @racket[_expr1] becomes a redex, its @tech{continuation} is the same
-as was the enclosing @racket[_expr2]'s @tech{continuation}.
+as the enclosing @racket[_expr2]'s @tech{continuation}.
```

Comment:
I think the “was” here is valuable: for a form to be in tail position, *`expr2`* must have disappeared by the time *`expr1`* becomes a redex, since otherwise *`expr2`* would be a part of *`expr1`*’s continuation. Using “was” emphasizes that *`expr2`* has been reduced away.

Reply:
Is it the case that _expr2_ must always have disappeared? If a function call is in tail position and that function always returns its first argument, isn't the first argument in tail position?

Reply:
No, in fact no argument is ever in tail position with respect to the function call.

Reply:
I’m not completely sure what you’re asking. Let’s consider the following expression:

```racket
(+ 1 ((lambda (x) x) (+ 2 3)))
```

I see two interpretations of your question:

  1. You are asking if `x` is in tail position with respect to the surrounding application. In that case, the answer is **yes**. When the application becomes the redex, it is `((lambda (x) x) 3)`, and its continuation is `(+ [] 1)`. After a step of evaluation, the expression will be reduced to `(+ xloc 1)`, where `xloc` is a location with the value `3`. Therefore, the new redex is `xloc`, and its continuation is still `(+ [] 1)`, which is the same as the previous redex’s continuation, so `x` is in tail position with respect to the surrounding application.

  2. You are asking if `(+ 2 3)` is in tail position with respect to the surrounding application. In that case, the answer is **no**. When `(+ 2 3)` becomes the redex, its continuation is `(+ ((lambda (x) x) []) 1)`, which is larger than the continuation of the eventual application, `(+ [] 1)`.

In either case, what I said still holds: if an expression is in tail position, the surrounding expression will have disappeared by the time it becomes a redex.

Reply:
I think my confusion has to do with when expressions are evaluated. For example, I think of the arm of an `if` as being evaluated before it is given to the continuation of the `if`. So the surrounding `if` expression is still there. But I need to think of the expression as given to the continuation and then evaluated, so that the `if` is gone when the expression is the redex.